### ISSUE TITLE
[BUG] "keyInfo is not in PEM format or in base64 format" when upgrading to v5

### ISSUE THREAD

**Issue by forty** [node-saml/node-saml#361]
Because of this commit https://github.com/node-saml/node-saml/commit/8920013542bca7337a1330e8e73ceb7263114ee8 some of our certificates which contained spaces caused the error `TypeError: keyInfo is not in PEM format or in base64 format`, while it was working fine in previous version. 

As this breaking change was not explicitly stated in the changelog, I figured out that I'd log this bug in case someone else searches for it. The solution is to clean up the base64 of `idpCert` before passing it to node-saml (so no spaces, correct "=" padding, etc) - I don't know if node-saml should defensively perform the clean up too, up to you (let me know if you'd like a PR for this, happy to do it).

**Comment by forty** [comment]
note that I choose the bug category, though it isn't really a bug,  because I did not know what else to choose - since the breaking change happened in a major release, this is mostly for documentation purpose.

**Comment by cjbarth** [comment]
I'd be happy to see what a PR like this would look like. Were the spaces something that were generated by some tool you used or were they are artifact of some other processing you did with them?

**Comment by forty** [comment]
So I don't really know how those spaces appeared tbh, the certificates are provisioned by our customers and it seems some of them have somehow the end of line of the base64 turned to spaces, and we did not validate enough so we accepted those certificates and stored them as is (which worked fine until the update, so we never realized).

For the "fix" I don't imagine anything too fancy one of:
> keyData = keyData.replace(/ /g, ''); // drop only spaces
> keyData = keyData.replace(/\s/g, ''); // drop all white-spaces
> keyData = keyData.replace(/[^A-Za-z0-9\+\/=]/g,''); // drop all non base64 character

One potential issue (we did not encounter, but I think could happen) would also be missing padding, which probably used to work, but would be blocked by the new stricter regexp. I don't know if it would be a good idea to handle that too.

**Comment by cjbarth** [comment]
Wouldn't that affect the header line as well?

**Comment by forty** [comment]
No because the failing assertion is after checking for PEM headers https://github.com/node-saml/node-saml/blob/72a39dcf46a93a8679123db78b22ff82d6dd4096/src/crypto.ts#L63 so that's were I would do the replace

Our certificate are in base 64 without the PEM headers. When there are PEM header, it seems that the regexp doesn't validate the base 64 at all (.*) so it would also have avoided us the issue

**Comment by msavolainen-gofore** [comment]
We had the same problem. Our key files had Bag attributes and Key attributes blocks before begin private key line. Removing those lines fixed the issue.

**Comment by cjbarth** [comment]
Thank you @srd90 for this research. It does seem that, while we could allow for more lenient whitespace handing, it cautions that this "may run the risk of accepting text that was not intended to be accepted in the first place". However, "the choice of parsing strategy depends on the context of use"; we can pick and still be compliant. It seems we have picked, and we are compliant, and any cert used which has extraneous whitespace is actually not from a compliant generator.

The quote in that Stack Overflow answer where it says "Furthermore, parsers SHOULD ignore whitespace and other non-base64 characters" is referring to data before the encapsulation boundaries.

> Data before the encapsulation boundaries are
   permitted, and parsers MUST NOT malfunction when processing such
   data.  Furthermore, parsers SHOULD ignore whitespace and other non-
   base64 characters and MUST handle different newline conventions.

Unless I'm missing something, the approach currently implemented is the most correct one and there isn't a compelling reason to change.

Here are some relevant snippets:

> New implementations SHOULD emit the strict format (Figure 3)
   specified above.  The choice of parsing strategy depends on the
   context of use.

> Most extant parsers ignore blanks at the ends of lines; blanks at the
   beginnings of lines or in the middle of the base64-encoded data are
   far less compatible.  These observations are codified in Figure 1.
   The most lax parser implementations are not line-oriented at all and
   will accept any mixture of whitespace outside of the encapsulation
   boundaries (see Figure 2).  Such lax parsing may run the risk of
   accepting text that was not intended to be accepted in the first
   place (e.g., because the text was a snippet or sample).

> Generators MUST wrap the base64-encoded lines so that each line
   consists of exactly 64 characters except for the final line, which
   will encode the remainder of the data (within the 64-character line
   boundary), and they MUST NOT emit extraneous whitespace.  Parsers MAY
   handle other line sizes.  These requirements are consistent with PEM
   [[RFC1421](https://www.rfc-editor.org/rfc/rfc1421)].

**Comment by KiwiKilian** [comment]
While I appreciate the strict format, it would be really helpful to have an error message which actually tells, which key/cert is falsy.

**Comment by cjbarth** [comment]
That seems very reasonable @KiwiKilian . Can you put up a PR for that?

**Comment by robinelvin** [comment]
I'm not sure I understand the problem. My application pulls the settings from the database including the IdP-supplied certificate stored as a text field. Upon upgrading to v5 I'm now getting this error message. Here's an example of the cert I'm using which is now rejected unless I remove all line breaks in the base64 string:

```
-----BEGIN CERTIFICATE-----
MIIDXTCCAkWgAwIBAgIJALmVVuDWu4NYMA0GCSqGSIb3DQEBCwUAMEUxCzAJBgNV
BAYTAkFVMRMwEQYDVQQIDApTb21lLVN0YXRlMSEwHwYDVQQKDBhJbnRlcm5ldCBX
<-- more base64 encoded lines here -->
99CJ25RkAcSobWNF5zD0O6lgOo3cEdB/ksCq3hmtlC/DlLZ/D8CJ+7VuZnS1rR2n
aQ==
-----END CERTIFICATE-----
```

This format is widely used so I don't understand the need to enforce the strict certificate parsing.

**Comment by cksachdev** [comment]
I was facing similar issue after upgrading to v5.0.0. Solution was to update private key to start with 
```
-----BEGIN PRIVATE KEY-----
```

My private key file was having extra properties e.g.
```
Private-Key: (2048 bit, 2 primes)
modulus:
publicExponent:
privateExponent:
prime1:
prime2:
exponent1:
exponent2:
coefficient:
```
I deleted all the above sections and then it worked fine.  Below is an example of how the content of files are loaded in nodejs:

```
  decryptionKey = fs.readFileSync(path.resolve("./privkey.pem"), "utf8"),
  idpCert = fs.readFileSync(path.resolve("./idp.crt"), "utf8");
```
Hope this helps someone landing on this issue.

**Comment by GregdTd** [comment]
I encountered the same issue today. I sanitized my certificate and removed all spaces as @forty suggested.: https://github.com/node-saml/node-saml/issues/361#issuecomment-2025122452

**Comment by nnmn** [comment]
Just in case someone else will have the same problem: If you provide the IDP certificate in base64 format, the base64 check fails if you have linefeed at the end of the last line. Files encoded at least with my Ubuntu's base64 tool have \n at end of each line, including the last.

**Comment by mikkosuonio** [comment]
I wrote tests demostrating the case reported by @nnmn above. It seems to handle certificates in Base64 format with a trailing newline differently.

**Comment by cjbarth** [comment]
Since it looks like this was resolved by merging in #373, I'll close this. If there is some part of this that needs further attention, please open another issue, or, even better, open a PR with tests so we can discuss actual code.

**Comment by mikkosuonio** [comment]
I added a PR. I added more direct tests, which show that the problem still exists (previous tests were mislead probably by the changed message in the exception).
